lxd/device/cdi/spec.go

```diff
@@ -0,0 +1,116 @@
+
+	"github.com/canonical/lxd/lxd/instance"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+const (
+	// defaultNvidiaTegraMountSpecPath is default location of CSV files that define the modifications required to the OCI spec.
+	defaultNvidiaTegraMountSpecPath = "/etc/nvidia-container-runtime/host-files-for-container.d"
+)
+
+// defaultNvidiaTegraCSVFiles returns the default CSV files for the Nvidia Tegra platform.
+func defaultNvidiaTegraCSVFiles(rootPath string) []string {
+	files := []string{
+		"devices.csv",
+		"drivers.csv",
+		"l4t.csv",
+	}
+
+	paths := make([]string, 0, len(files))
+	for _, file := range files {
+		paths = append(paths, filepath.Join(rootPath, defaultNvidiaTegraMountSpecPath, file))
+	}
+
+	return paths
+}
+
+// generateNvidiaSpec generates a CDI spec for an Nvidia vendor.
+func generateNvidiaSpec(cdiID ID, inst instance.Instance, l logger.Logger) (*specs.Spec, error) {
+	mode := nvcdi.ModeAuto
+	if cdiID.Class == IGPU {
+		mode = nvcdi.ModeCSV
+	}
+
+	indexDeviceNamer, err := nvcdi.NewDeviceNamer(nvcdi.DeviceNameStrategyIndex)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create device namer with index strategy: %v", err)
+	}
+
+	uuidDeviceNamer, err := nvcdi.NewDeviceNamer(nvcdi.DeviceNameStrategyUUID)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create device namer with uuid strategy: %v", err)
+	}
+
+	nvidiaCTKPath, err := exec.LookPath("nvidia-ctk")
+	if err != nil {
+		return nil, fmt.Errorf("Failed to find the nvidia-ctk binary: %v", err)
+	}
+
+	rootPath := ""
+	if shared.InSnap() {
+		rootPath = "/var/lib/snapd/hostfs"
+	}
+
+	cdilib, err := nvcdi.New(
+		nvcdi.WithDeviceNamers(indexDeviceNamer, uuidDeviceNamer),
+		nvcdi.WithLogger(NewCDILogger(l)),
+		nvcdi.WithDriverRoot(rootPath),
+		nvcdi.WithDevRoot(rootPath),
+		nvcdi.WithNVIDIACDIHookPath(nvidiaCTKPath),
+		nvcdi.WithMode(mode),
+		nvcdi.WithCSVFiles(defaultNvidiaTegraCSVFiles(rootPath)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create CDI library: %v", err)
+	}
+
+	specIface, err := cdilib.GetSpec()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get CDI spec interface: %v", err)
+	}
+
+	spec := specIface.Raw()
+	if spec == nil {
+		return nil, fmt.Errorf("CDI spec is nil")
+	}
+
+	// The spec definition can be quite large so we log it to a file.
+	specPath := filepath.Join(inst.LogPath(), fmt.Sprintf("nvidia_cdi_spec.%s.log", strings.ReplaceAll(cdiID.String(), "/", "_")))
+	specFile, err := os.Create(specPath)
+	if err != nil {
+		logger.Warn("Failed to create a log file to hold a CDI spec", logger.Ctx{"specPath": specPath, "error": err})
+		return spec, nil
+	}
+
+	defer specFile.Close()
+
+	_, err = specFile.WriteString(logger.Pretty(spec))
+	if err != nil {
+		return nil, fmt.Errorf("Failed to write spec to %q: %v", specPath, err)
+	}
+
+	logger.Debug("CDI spec has been successfully generated", logger.Ctx{"specPath": specPath})
```

Comment:
Please use the logger with context mentioned here https://github.com/canonical/lxd/pull/13562/commits/c6c141ac2966c9f525d72cbbe4587ce8baf88106#r1706693630

Reply:
fixed